### PR TITLE
Set users with incorrectly populated displayName as private

### PR DESCRIPTION
**Pull request by sallakarppinen**
This pull request provides a fix for issue #1284

### CODE REVIEW

File: node_modules/oae-authentication/lib/strategies/shibboleth/init.js [hunk 1]
```diff
@@ -84,6 +84,11 @@ module.exports = function(config) {
                 'authoritative': true
             };
 
+            // Set users whose name resembles a Shibboleth identifier as private (eg. starts with http://)
+            if (displayName.indexOf('http://') === 0 || displayName.indexOf('shibboleth!') != -1) {
```
Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/lib/strategies/shibboleth/init.js]: We should also catch display names that are email addresses (`@`)

File: node_modules/oae-authentication/lib/strategies/shibboleth/init.js [hunk 2]
```diff
@@ -84,6 +84,11 @@ module.exports = function(config) {
...
+                opts.visibility ='private';
+            } 
+
```
Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/lib/strategies/shibboleth/init.js]: When the user has an invalid display name / email address, the `editprofile` widget will be displayed in the UI. At that point, the user's account visibility should not show as private, but as the default profile visibility for that tenant, so once they save they get the default visibility (unless they change it).

File: node_modules/oae-authentication/lib/strategies/shibboleth/init.js (hunk 1 above)
Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/lib/strategies/shibboleth/init.js]: This should probably catch `https://` as well

File: node_modules/oae-authentication/tests/test-external-strategies.js [hunk 3]
```diff
@@ -985,6 +985,53 @@ describe('Authentication', function() {
...
+                        'shib-session-id': Math.random(),
+
+                        // Fake some data about the IdP
+                        'persistent-id': 'https://idp.example.com/shibboleth#https://sp.example.com/shibboleth#' + Math.random(),
+                        'identityProvider': 'https://idp.example.com/shibboleth',
+                        'affiliation': 'Digital Services',
+                        'unscopedAffiliation': 'OAE Team',
+
+                        // Generate an external id
+                        'remote_user': 'shibboleth!' + Math.random(),
```
Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/tests/test-external-strategies.js]: Can we extend this to test all possible cases (`http`, `https`, `@`) ?

**Comment by nicolaasmatthijs** [comment]
Finished review. Back to @sallakarppinen for follow-up

File: node_modules/oae-authentication/lib/strategies/shibboleth/init.js (hunk 2 above)
Comment (sallakarppinen) [inline node_modules/oae-authentication/lib/strategies/shibboleth/init.js]: If I understand correctly, that widget is in 3akai-ux, so do I need to make a pull request there if I need to make changes to it?

Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/lib/strategies/shibboleth/init.js]: That's correct

Comment (sallakarppinen) [inline node_modules/oae-authentication/lib/strategies/shibboleth/init.js]: Cool - I'll have a look at how to do this and will then create a new pull request over there. Meanwhile the issues you've raised here should now all be fixed (I've extended the check to look for emails and added test cases).

File: node_modules/oae-authentication/lib/strategies/shibboleth/init.js [hunk 4]
```diff
@@ -84,6 +84,12 @@ module.exports = function(config) {
                 'authoritative': true
             };
 
+            var invalid = /https?:\/\/|shibboleth!|@/i;
+            // Set users whose name resembles a Shibboleth identifier as private (eg. starts with http://)
+            if (invalid.test(displayName)) {
+                opts.visibility ='private';
```
Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/lib/strategies/shibboleth/init.js]: Should be using the constant `AuthzConstants.visibility.PRIVATE`

File: node_modules/oae-authentication/tests/test-external-strategies.js [hunk 5]
```diff
@@ -985,6 +985,147 @@ describe('Authentication', function() {
         });
 
         /**
+         * Test that verifies that the user will be set to private if displayName resembles a shibboleth identifier
```
Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/tests/test-external-strategies.js]: As we're essentially using the same test 3 times, it'd be better to abstract a function that does the majority of the checks and we can just pass a remote user into. We can then execute it 4 times (for the `htttp://` case as well).

**Comment by nicolaasmatthijs** [comment]
Finished review. Back to @sallakarppinen for follow-up

File: node_modules/oae-authentication/tests/test-external-strategies.js [hunk 6]
```diff
@@ -984,6 +984,67 @@ describe('Authentication', function() {
             }, callback);
         });
 
+        var verifyInvalidDisplayNameMakesProfilePrivate = function(remoteUser, callback) {
```
Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/tests/test-external-strategies.js]: We require all functions to have JSDocs (including parameter documentation).

**Comment by nicolaasmatthijs** [comment]
Left one documentation related comment. Back to @sallakarppinen

File: node_modules/oae-authentication/lib/strategies/shibboleth/init.js [hunk 7]
```diff
@@ -29,6 +29,7 @@ var AuthenticationUtil = require('oae-authentication/lib/util');
 var ShibbolethAPI = require('./api');
 var ShibbolethStrategy = require('./strategy');
 
+var AuthzConstants = require('oae-authz/lib/constants').AuthzConstants;
```
Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/lib/strategies/shibboleth/init.js]: We follow the following ordering when requiring dependencies:

1. All external dependencies
2. All OAE dependencies from different modules
3. All OAE dependencies from the current module

Therefore, `AuthzConstants` should be added to the `var Cassandra = ...` block.

Note that we order the requires alphabetically within their block.

File: node_modules/oae-authentication/tests/test-external-strategies.js [hunk 8]
```diff
@@ -985,6 +985,74 @@ describe('Authentication', function() {
         });
 
         /**
+         * Test that verifies that the user visibility will be set to private when remote_user is used to populate
```
Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/tests/test-external-strategies.js]: As this isn't an actual test, it should be documented as something along the lines of `Test util that ...`

File: node_modules/oae-authentication/tests/test-external-strategies.js [hunk 9]
```diff
@@ -985,6 +985,74 @@ describe('Authentication', function() {
...
+                    // Fake a session ID to log in
+                    'shib-session-id': Math.random(),
+
+                    // Fake some data about the IdP
+                    'persistent-id': 'https://idp.example.com/shibboleth#https://sp.example.com/shibboleth#' + Math.random(),
+                    'identityProvider': 'https://idp.example.com/shibboleth',
+                    'affiliation': 'Digital Services',
+                    'unscopedAffiliation': 'OAE Team',
+
+                    // Generate an external id
```
Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/tests/test-external-strategies.js]: This comment should be updated to reflect what's happening

File: node_modules/oae-authentication/tests/test-external-strategies.js [hunk 10]
```diff
@@ -985,6 +985,74 @@ describe('Authentication', function() {
...
+                        assert.strictEqual(me.visibility, 'private');
+                        callback();
+                    });
+                });
+            });
+        }
+
+        /**
+         * Test that verifies that the user visibility will be set to private if displayName resembles a shibboleth identifier,
+         * URL or email address.
```
Comment (nicolaasmatthijs) [inline node_modules/oae-authentication/tests/test-external-strategies.js]: The dot at the end can be removed

**Comment by nicolaasmatthijs** [comment]
Added some minor review to help highlight some of our documentation guidelines. Back to @sallakarppinen for follow-up

**Comment by nicolaasmatthijs** [comment]
Looks good. Merged.

**Pull request merged**